rfd/0005-kubernetes-service.md

```diff
@@ -0,0 +1,514 @@
+  kube_listen_addr: 0.0.0.0:3026
+
+kubernetes_service:
+  enabled: yes
+```
+
+```sh
+$ tsh kube clusters
+Cluster Name       Status
+-------------      ------
+example.com        online
+```
+
+Note: `kubernetes_service` doesn't explicitly set `kube_cluster_name`, so it
+defaults to the Teleport `cluster_name`.
+
+Next, user connects a new k8s cluster to the existing auth server:
+
+`teleport-other.yaml`:
+```yaml
+teleport:
+  auth_servers: auth.example.com:3025
+
+kubernetes_service:
+  enabled: yes
+  kube_cluster_name: other.example.com
+  # Note: public_addr/listen_addr are needed for the proxy_service to connect
+  # to this kubernetes_service
+  public_addr: other.example.com:3026
+  listen_addr: 0.0.0.0:3026
+```
+
+```sh
+$ tsh kube clusters
+Cluster Name       Status
+-------------      ------
+example.com        online
+other.example.com  online
+```
+
+Next, user connects a third k8s cluster over a reverse tunnel to the proxy:
+
+`teleport-third.yaml`:
+```yaml
+teleport:
+  auth_servers: proxy.example.com:3080
+
+kubernetes_service:
+  enabled: yes
+  kube_cluster_name: third.example.com
+  # Note: public_addr/listen_addr are NOT needed because this
+  # kubernetes_service connects to proxy_service over a reverse tunnel (via
+  # auth_servers address above).
+```
+
+```sh
+$ tsh kube clusters
+Cluster Name       Status
+-------------      ------
+example.com        online
+other.example.com  online
+third.example.com  online
+```
+
+#### Scenario 2
+
+A single central Teleport instance connected to multiple k8s clusters.
+
+`teleport.yaml`:
+```yaml
+proxy_service:
+  public_addr: proxy.example.com:3080
+  kube_listen_addr: 0.0.0.0:3026
+
+kubernetes_service:
+  enabled: yes
+  kubeconfig_file: kubeconfig
+```
+
+`kubeconfig`:
+```yaml
+apiVersion: v1
+kind: Config
+clusters:
+- cluster:
+    server: https://a.example.com
+  name: a-cluster
+- cluster:
+    server: https://b.example.com
+  name: b-cluster
+contexts:
+- context:
+    cluster: a-cluster
+    user: shared-creds
+  name: a
+- context:
+    cluster: b-cluster
+    user: shared-creds
+  name: b
+current-context: a
+users:
+- name: shared-creds
+  user:
+    client-key-data: "..."
+    client-certificate-data: "..."
+```
+
+```sh
+$ tsh kube clusters
+Cluster Name       Status
+-------------      ------
+a                  online
+b                  online
+```
+
+#### Scenario 3
+
+A single central Teleport proxy acting as "gateway". Multiple k8s clusters
+connect to it over reverse tunnels.
+
+`teleport-proxy.yaml`:
+```yaml
+proxy_service:
+  public_addr: proxy.example.com:3080
+  kube_listen_addr: 0.0.0.0:3026
+```
+
+`teleport-a.yaml`:
+```yaml
+teleport:
+  auth_servers: proxy.example.com:3080
+
+kubernetes_service:
+  enabled: yes
+  kube_cluster_name: a.example.com
+```
+
+`teleport-b.yaml`:
+```yaml
+teleport:
+  auth_servers: proxy.example.com:3080
+
+kubernetes_service:
+  enabled: yes
+  kube_cluster_name: b.example.com
+```
+
+```sh
+$ tsh kube clusters
```

Comment:
Do end users need to do anything else with `kubectl` for multiple cluster? or does it just work out of the box? https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/

Reply:
`tsh login` will configure all known clusters as `kubectl` contexts.
to switch between them, end users have to call either `tsh kube login $cluster_name` or `kubectl config use-context $cluster_name`.